[1단계 - 상품 관리 기능] 도이(유도영) 미션 제출합니다.

yoondgu opened the pull request:
안녕하세요 빙봉 🐘  리뷰이 도이입니다!

이렇게 코드 리뷰를 기회로 만나뵙게 되어 반갑습니다. 따끔한 피드백 부탁드려요!! 
제가 부족한 점이 많겠지만 같이 재밌는 대화 많이 나눠보고 싶어요. ㅎㅎ 감사합니다!

## 실행환경 설명

어플리케이션 실행 시 `resources/schema.sql`의 DDL이 실행됩니다.
DB는 임베디드 H2를 사용하였고, 실제 서비스와 테스트 전용 DB를 분리하였습니다.

먼저 인메모리가 아닌 방식으로 하여 좀 더 실제 서비스 환경과 유사한 상태에서 구현, 테스트를 진행하고 싶었고,
H2에서도 어플리케이션 종료와 관계 없이 유지되는 형식이 가능하다고 해서 사용해보았는데요.

`application.properties` 파일에 설정한 대로 로컬의 home 경로에 H2 DB 파일이 생성된다는 점 양해 부탁드립니다. 🥲 
만약 이 방법이 적절치 않거나 리뷰 진행에 불편함이 있으시다면 MySQL DB로 변경하도록 하겠습니다!

## 고민한 점

### 서비스 계층 정의하지 않은 이유

- 특별한 비즈니스 로직 없이 DB 액세스만 진행되기 때문에 서비스 계층은 꼭 필요하지 않은 상태라고 판단했습니다.
- 다만 update, delete의 경우 해당하는 id가 존재하는지 확인하는 검증 로직(`productDao.findById` 의 반환 값 null 여부 확인)을 추가해야 한다고 생각합니다. 그렇다면 “검증 로직 + DB 액세스“ 두 가지 일을 하게 되므로 이를 서비스 계층으로 분리해주면 좋을 것 같습니다.
하지만 이 부분은 페어와의 의견 차이가 있었기에 보류하였고, 리팩터링 과정에서 반영하려고 합니다!
- 마찬가지로 현재 도메인 객체인 Product는 그대로 DB 액세스에 사용합니다. 상태 변경의 위험이 없기 때문에 Dao 전용 Entity를 만들고 변환해 DB 액세스하는 작업은 필요 없다고 생각했습니다.

### Controller 분리 기준

- 주어진 html, js 파일을 봤을 때, Model에 정보를 담고 뷰를 반환해주어야 하는 요청과 raw한 데이터 그 자체를 반환해주어야 하는 요청이 있었습니다. 따라서 이를 기준으로 컨트롤러를 두 개로 분리했습니다.
- post 요청의 경우 `CartController` 에서 redirect 시키도록 PRG 패턴을 적용해보려고 했지만, 프론트엔드 코드에서는
1. C, U, D에 대해서 비동기로 요청을 보냄
2. 요청에 대해 정상적으로 응답을 받으면 화면을 새로 고침시킴
    
    위와 같이 처리하고 있었습니다. 이에 대해 프론트엔드 코드를 수정하기보다는 주어진 비동기 요청에 맞게 RestController에서 응답하는 것이 미션에서 요구하는 바라고 이해했습니다. 
    

### Id 필요 유무에 따라 DTO 클래스 분리

- create 시에는 id를 받지 않고, update 시에는 id를 받아야 합니다. 하나의 DTO를 사용하고 id 필드에 대해서는 null을 허용해주는 것이 null safe하지 못한 상황이라고 생각해, 각각 다른 DTO를 사용하도록 했습니다.
- 하지만 이런 방식으로 매 요청마다 다른 DTO를 생성하게 된다면 관리하기 어렵고, 남용의 여지가 있을 것 같아 이를 Inner class로 정의해보았습니다.
❓ 혹시 이는 일반적이지 않은 방식일까요? 또는 이 방식에 대해 제가 생각하지 못한 단점이 있을까요?

### 검증 로직의 책임 분리

- 계층 별로 검증을 해주는 것은 좋다고 생각하지만, DTO에 중복되는 검증 어노테이션이 너무 많아서 이를 제거해보고 싶었습니다.
- DTO는 request body의 값을 잘 받아왔는지에 대한 검증만 하고 (not null),
그 외의 검증 로직(문자열 길이, 금액 범위 등)은 도메인 영역이 책임지도록 분리해보았습니다.

### 그 외 질문드리고 싶은 점

❓ 기본 생성자, getter, equals & hashCode 정도는 Lombok 어노테이션을 사용해 정의하는 것이 가독성, 유지보수성 측면에서 모두 좋다고 생각해 사용하고 있습니다. 그런데 이 어노테이션들을 정의하는 순서에도 혹시 권장되는 컨벤션이 있을까요? 

❓ 요청 URI의 경우 클라이언트 입장에서 이해하기 쉽도록 프론트엔드 코드의 각 요청 관련 함수에서 쓴 단어들(add,edit,delete)로 구성했습니다. 하지만 REST API는 URI에서는 자원만 명사로 명시하고, HTTP Method로 행위를 표현한다고 알고 있습니다. 현재 코드 또한 HTTP Method만으로도 행위를 구분할 수 있기 때문에 추가로 붙여준 path가 불필요한가? 생각이 듭니다. (하지만 아직 REST API에 대해 잘 알지 않아서 따로 수정은 안했습니다..!) 빙봉은 이에 대해 어떻게 생각하시는지 궁금해요!

yoondgu committed:
docs: 미션 요구사항 작성

yoondgu committed:
feat: 도메인 객체 구현

yoondgu committed:
feat: 상품 목록 페이지 연동

yoondgu committed:
feat: 데이터 베이스 설계

yoondgu committed:
docs: API 요청 스펙 작성

yoondgu committed:
feat: DAO 구현 완료

yoondgu committed:
feat: 관리자 도구 페이지 url 매핑

yoondgu committed:
feat: 상품 추가  기능 구현

yoondgu committed:
feat: 상품 수정, 삭제 기능 구현

yoondgu committed:
feat: Request 예외 처리 기능 구현

yoondgu committed:
refactor: 각 요청에 맞는 Dto로 분리

yoondgu committed:
test: 통합 테스트 작성

yoondgu committed:
style: 불필요한 주석 삭제

yoondgu committed:
refactor: 변수 이름 변경

yoondgu committed:
chore: TABLE 생성 DDL 초기 실행 설정

yoondgu committed:
refactor: 사용되지 않는 필드 제거

yoondgu committed:
refactor: 사용되지 않는 클래스 제거

yoondgu committed:
refactor: 검증 책임 domain 객체로 이동

yoondgu committed:
refactor: 테스트 전용 DB 초기화 실행 어노테이션 적용

yoondgu committed:
refactor: 도메인 검증로직 관련 상수화

yoondgu committed:
style: 코드 줄바꿈 방식 통일

yoondgu committed:
test: 테스트 누락 displayName 추가

aegis1920 commented on README.md:
> 특별한 비즈니스 로직 없이 DB 액세스만 진행되기 때문에 서비스 계층은 꼭 필요하지 않은 상태라고 판단했습니다. 다만 update, delete의 경우 해당하는 id가 존재하는지 확인하는 검증 로직(productDao.findById 의 반환 값 null 여부 확인)을 추가해야 한다고 생각합니다. 그렇다면 “검증 로직 + DB 액세스“ 두 가지 일을 하게 되므로 이를 서비스 계층으로 분리해주면 좋을 것 같습니다. 하지만 이 부분은 페어와의 의견 차이가 있었기에 보류하였고, 리팩터링 과정에서 반영하려고 합니다!

아.. 이런 관점 너무 좋아요. 도이 최고 👍 👍 리팩터링 진행해보시죠! (서비스 레이어 추가하실 때 서비스 레이어 테스트도 추가 부탁드립니다!)

> 주어진 html, js 파일을 봤을 때, Model에 정보를 담고 뷰를 반환해주어야 하는 요청과 raw한 데이터 그 자체를 반환해주어야 하는 요청이 있었습니다. 따라서 이를 기준으로 컨트롤러를 두 개로 분리했습니다.

👍 

> 혹시 이는 일반적이지 않은 방식일까요? 또는 이 방식에 대해 제가 생각하지 못한 단점이 있을까요?

음.. 일반적이지 않은 방식이라기보단 Controller 클래스인데 Controller의 역할을 하고 있는지 생각해보면 좋을 것 같습니다!
요청 클래스로 분리되어야하지 않을까요? e.g. Request.ProductAdd

> DTO는 request body의 값을 잘 받아왔는지에 대한 검증만 하고 (not null),
그 외의 검증 로직(문자열 길이, 금액 범위 등)은 도메인 영역이 책임지도록 분리해보았습니다.

👍 

> 기본 생성자, getter, equals & hashCode 정도는 Lombok 어노테이션을 사용해 정의하는 것이 가독성, 유지보수성 측면에서 모두 좋다고 생각해 사용하고 있습니다. 그런데 이 어노테이션들을 정의하는 순서에도 혹시 권장되는 컨벤션이 있을까요?

권장되는 컨벤션이 있는지 모르겠네요? 음.. 정의하는 순서에 따라 기능이 달라지는 게 아니기 때문에 권장되는 컨벤션은 없다고 생각합니다! 저희 팀에서도 롬복 순서 관련된 컨벤션은 없습니다.

> 요청 URI의 경우 클라이언트 입장에서 이해하기 쉽도록 프론트엔드 코드의 각 요청 관련 함수에서 쓴 단어들(add,edit,delete)로 구성했습니다. 하지만 REST API는 URI에서는 자원만 명사로 명시하고, HTTP Method로 행위를 표현한다고 알고 있습니다. 현재 코드 또한 HTTP Method만으로도 행위를 구분할 수 있기 때문에 추가로 붙여준 path가 불필요한가? 생각이 듭니다. (하지만 아직 REST API에 대해 잘 알지 않아서 따로 수정은 안했습니다..!) 빙봉은 이에 대해 어떻게 생각하시는지 궁금해요!

저도 도이와 같은 생각입니다! HTTP Method만으로도 행위를 구분할 수 있기 때문에 추가로 붙여준 path가 불필요해보여요. 다만 복잡한 요청이 들어올 때 말씀해주신 단어를 붙이기도 합니다. 이런 부분은 추후 구현하시면서 직접 느끼실거예요!

aegis1920 commented on src/main/java/cart/controller/CartController.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @GetMapping(path = "/")
+    public String home(Model model) {
```
- Model이 반환되는 과정을 말씀해주실 수 있을까요?
- Model이 아닌 ModelAndView를 써보는 건 어떨까요?

yoondgu replied:
현재 코드와 같이 String으로 view 이름을 반환하면,
HandlerAdapter가 반환값을 분석하여(ViewResolver) 적절한 View를 찾은 뒤 ModelAndView 객체를 만들어서 DispatchServlet에게 제공하는 것으로 알고 있습니다.

- Model을 사용하는 방식의 장점
    - ModelAndView 객체를 생성해 반환할 필요 없다
    - 모델 속성에 대해서 더 유연하게 조작할 수 있다 (e.g. mergeAttributes 메서드)
- ModelAndView를 사용하는 방식의 장점
    - Model과 View를 일괄적으로 관리한다 (엉뚱한 뷰 이름을 반환한다든지 하는 위험이 적을 수 있을 것 같아요)
    - 핸들러 메서드의 시그니처가 간결하고, 의미가 명확하다
         - Model을 매개변수로 전달받지 않으므로 핸들러메서드의 매개변수로는 요청으로부터 전달받는 값만을 받고, 반환값으로는 ModelAndView를 반환하기 때문이다.
         
제가 정리해 본 장점은 위와 같은데요!
ModelAndView를 쓰면, 
- 핸들러 메서드 내부 코드에서 해당 객체를 생성해야 한다는 것 -> 개발자가 신경써야 한다는 점이 늘어난다고 볼 수 있지 않을까요?
- 어차피 HandlerAdapter에서 DispatcherServlet으로 반환할 때 ModelAndView 변환을 해주기 때문에 직접 할 필요성이 있는지 의문입니다.
빙봉은 이 점에 대해 어떻게 생각하시나요?!

aegis1920 replied:
- 뷰를 렌더링하는 과정에 대해 한 번 디버깅해보셨으면 좋겠다는 의미에서 질문드렸는데 너무 잘 알고 계시군요 👍 👍 
- 디버깅하면서 DispatcherServlet.doDispatch() -> ... -> RequestMappingHandlerAdapter.getModelAndView() 이쪽을 간단하게 보시면 좋을 것 같아요!

> 핸들러 메서드 내부 코드에서 해당 객체를 생성해야 한다는 것 -> 개발자가 신경써야 한다는 점이 늘어난다고 볼 수 있지 않을까요?

핸들러 메서드 내부 코드는 어떤 걸 의미하신 건지 말씀해주실 수 있을까요? RequestMappingHandlerAdapter.getModelAndView() 이쪽에서 ModelAndView를 생성하는 걸 말씀하신 걸까요? 만약 맞다면 Model 인터페이스를 사용하더라도 ModelAndView 객체를 생성합니다!

> 어차피 HandlerAdapter에서 DispatcherServlet으로 반환할 때 ModelAndView 변환을 해주기 때문에 직접 할 필요성이 있는지 의문입니다.

필요성이 있는 건 아닙니다! 제가 `Model이 아닌 ModelAndView를 써보는 건 어떨까요?`라고 질문드렸던 이유는 다양한 방법이 있으니 한 번 작성해보면 어떨까? 하는 취지였습니다ㅎㅎ Controller에서 Model, ModelAndView 뿐만 아니라 다양한 형태로 반환시켜줄 수 있는데요. 시간되시면 org.springframework.web.servlet.View 인터페이스를 상속한 여러 View를 살펴보셔도 도움이 될 것 같습니다 🙂 

아 개인적으로 저는 Model보단 ModelAndView가 좀 더 명시적인 코드라고 생각이 들긴 해서요. ModelAndView를 더 선호하긴 합니다!

yoondgu replied:
"내부 코드"는 위와 같이 개발자가 직접 객체 생성해야 하는 부분을 이야기하고 싶었습니다!
Model에 값을 담고, 뷰만 반환해주면 개발자가 직접 객체 생성할 필요 없이 설명해주신 것처럼 ModelAndView 객체를 생성하기 때문입니다.

```java
    @GetMapping(path = "/")
    public ModelAndView home() {
        // ModelAndView 객체를 직접 생성해서 반환
        return new ModelAndView("index", Map.of(
                "products", productService.findAll()
        ));
    }
```

덕분에 하나의 방법만 있는 것이 아니라는 것, 각 방법의 장단점에 대해 고민해볼 수 있었어요 ㅎㅎㅎ 감사합니다. 학습할 때 디버깅을 어디를 해야 할지 항상 막막한데, 살펴볼만한 지점들을 짚어주셔서 도움이 되었습니다!

저 역시 ModelAndView를 사용하면 더 명시적인 코드가 된다는 점에 공감합니다!
둘 다 장단점이 있다고 생각이 되지만, ModelAndView는 직접 제대로 써본 적이 없기 때문에 학습 목적 차원에서 이를 사용해보도록 하겠습니다 ㅎㅎ

aegis1920 commented on src/main/java/cart/controller/ProductController.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    @PostMapping(path = "/add")
+    public ResponseEntity<Void> create(@Valid @RequestBody ProductAddRequest productDto) {
```
자원을 생성할 땐 어떤 응답 코드를 내려보내야 할까요?

yoondgu replied:
자원을 생성할 경우 201 Created를 보내도록 수정하겠습니다!

yoondgu replied:
201 상태코드로 보낼 경우 Location에 생성한 자원의 uri를 담아 보내기도 한다는 것을 알았지만,
지금 당장은 해당 자원의 uri가 존재하지 않으므로 이를 생략했습니다.

aegis1920 replied:
근거가 있어서 좋습니다! 👍

aegis1920 commented on src/main/java/cart/controller/ProductController.java:
```diff
@@ -0,0 +1,120 @@
+        this.productDao = productDao;
+    }
+
+    @PostMapping(path = "/add")
```
개인적인 취향인데요. 저는 `@RequestMapping`을 없애고 중복이 되더라도 URL을 분리하지 않고 한 번에 쓰는 편입니다. (e.g. `@PostMapping("/product/add")`)
그래야 찾기 편하더라구요ㅎㅎ (물론 분리해도 shift + shift로 뜨는 창에서 `/url 검색할 URL` 이런 식으로 찾을 수 있긴 합니다)

aegis1920 commented on src/main/java/cart/controller/ProductController.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    @PutMapping(path = "/edit")
+    public ResponseEntity<Void> update(@Valid @RequestBody ProductUpdateRequest productDto) {
```
현재 수정버튼을 클릭하면 이미지가 undefined로 나오는 문제가 있네요? 어떤 것 때문일까요?
![image](https://user-images.githubusercontent.com/22311557/234617233-60c2cd12-b525-4f6d-b5af-4ad6a39bfc36.png)

추가한 이미지 링크는 `https://media.istockphoto.com/id/1187850123/ko/%EC%82%AC%EC%A7%84/%ED%97%AC%EB%A6%AC%EC%BD%A5%ED%84%B0.jpg?s=1024x1024&w=is&k=20&c=7l_JJ-nJ_elIPD0lK4c0c9NLZxvZ_ympvYGxa9bY0yo=` 입니다.

yoondgu replied:
DB에서 조회시 사용하는 도메인 객체 Product의 필드가 프론트의 필드명과 일치 하지 않아서 생기는 문제였습니다..!!
DTO에서는 @JsonProperty로 이를 매핑해주었는데 (javascript에서는 snake case, java에서는 carmel case를 쓰는 일관성을 유지하고 싶었어요..ㅎㅎ) 조회 시 사용하는 객체에서도 해줘야 한다는 것을 놓쳤네요.
알려주셔서 감사합니다!! 버그 해결하여 푸쉬하겠습니다 ㅎㅎ

aegis1920 commented on src/main/java/cart/controller/ProductController.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    @DeleteMapping("/delete/{id}")
+    public ResponseEntity<Void> delete(@PathVariable("id") @Positive Long id) {
+        productDao.deleteById(id);
+        return ResponseEntity.ok().build();
```
- 아무것도 반환하지 않는데 200을 반환하신 이유가 있을까요?
- 아무것도 반환하지 않는다면 어떤 응답 코드를 반환해야할까요?

yoondgu replied:
적절한 상태코드에 대해 잘 모르는 상태였어서 성공적인 응답이라는 의미만으로 200을 반환했는데, 
요청을 처리했지만 body를 반환할 필요 없는 경우 명시적으로 204 코드를 반환한다는 것을 알았습니다! 수정하겠습니다 ㅎㅎ

aegis1920 replied:
REST API 관련 몇 가지 링크 추천드려요! (REST API는 github이 원칙을 잘 지키고 있는 것으로 소문이 자자하니 github docs를 보시는 걸 추천드려요)

- https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28
- https://youtu.be/RP_f5dMoHFc

yoondgu replied:
참고 링크 추천해주셔서 감사합니다!! 😆
복수/단수 표현에 대해서도 고민이 있었는데
어떤 요청은 복수의 자원에 대한 것이고, 어떤 것은 단수의 자원에 대한 것일텐데
이 둘을 혼용해서 쓰는것은 혼란스러울 수 있으니 지양해야 하는 것 같고,
확장 가능한 API가 될 수 있도록 하려면 더 넓은 범위에서 이해할 수 있는 복수형을 쓰는 게 좋다고 생각이 됩니다..!!
따라서 uri의 단수를 복수로 변경하려고 합니다.

aegis1920 commented on src/main/java/cart/controller/ProductController.java:
```diff
@@ -0,0 +1,120 @@
+
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ErrorDto handleException(MethodArgumentNotValidException exception) {
```
- handleException을 하는 역할이 Controller의 역할은 아닌 것 같아요. 따로 빼는 건 어떨까요?
- `@RestControllerAdvice`도 활용해보세요!

yoondgu replied:
저도 빼는 게 더 좋다고 생각합니다. 반영할게요!

aegis1920 commented on src/main/java/cart/dao/ProductDao.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.List;
+import java.util.Optional;
+
+public interface ProductDao {
```
지금 상황에서는 Product 하나 밖에 없어서 상관없을 것 같긴한데요. 제네릭을 적용해봐도 좋을 것 같습니다 ([링크](https://www.baeldung.com/java-dao-pattern))

aegis1920 commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,8 @@
+# H2
+spring.h2.console.enabled=true
+# Datasource
+spring.datasource.url=jdbc:h2:file:~/shopping-cart
```
> 먼저 인메모리가 아닌 방식으로 하여 좀 더 실제 서비스 환경과 유사한 상태에서 구현, 테스트를 진행하고 싶었고,
H2에서도 어플리케이션 종료와 관계 없이 유지되는 형식이 가능하다고 해서 사용해보았는데요.

👍 

> application.properties 파일에 설정한 대로 로컬의 home 경로에 H2 DB 파일이 생성된다는 점 양해 부탁드립니다. 🥲 
만약 이 방법이 적절치 않거나 리뷰 진행에 불편함이 있으시다면 MySQL DB로 변경하도록 하겠습니다!

home 경로에 생성하지 않고 프로젝트 폴더 내에 생성해도 됐을 것 같아요! 물론 도이가 git push 하실 땐 도이의 파일이 올라가지 않도록 git ignore 해줘야겠네요 🙂 
```suggestion
spring.datasource.url=jdbc:h2:file:./db/shopping-cart
```

질문이 하나 있는데요. MySQL DB로 변경하면 어떤 게 달라지나요..?

yoondgu replied:
아 경로를 프로젝트 폴더 내로 설정해주면 되는 문제였네요 ㅎㅎㅎ 감사합니다 반영하겠습니다!!
MySQL 같은 다른 DB는 해당 프로그램 관련 디렉토리에 저장되도록 기본 설정이 되어있어서 더 낫겠지? 라고 생각했는데 
사실 파일 저장 위치는 어떤 벤더사이든 설정할 수 있는 부분일 것 같아서 제가 잘못 말씀 드린 것 같아요..!!

aegis1920 commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE IF NOT EXISTS products (
+    id BIGINT NOT NULL AUTO_INCREMENT,
```
autoincrement의 경우, unsigned를 사용하면 범위를 더 넓힐 수 있습니다.

yoondgu replied:
정수 타입에 unsigned를 붙이면 음수 값을 허용하지 않고, 양수값의 범위를 더 넓혀준다는 것을 처음 알았습니다! 감사합니다 반영하겠습니다 😀
그리고 H2 SQL은 unsigned를 지원하지 않는다는 것을 알아서, MySQL 문법으로 동작하도록 applicaition.properties 파일에서 H2의 Mode를 설정하였습니다.

aegis1920 reviewed (changes requested):
안녕하세요 도이! 빙봉입니다 🙂 

질문에 대한 답변 및 코멘트 남겼으니 확인해주세요~

궁금하신 점은 편하게 DM 주세요!

yoondgu committed:
refactor: 테스트 전용 DB 초기화 자체 설정, 어노테이션 삭제

yoondgu committed:
docs: 코드리뷰 피드백 및 리팩터링 목록 작성

yoondgu committed:
refactor: 사용하지 않는 변수 할당 코드 삭제

yoondgu committed:
fix: 수정 모달 폼 이미지 URL undefined 버그 해결

- 데이터 조회 자바 객체와 js 코드 필드 매핑 문제로 발생한 버그임.

yoondgu committed:
refactor: @RestController의 반환 타입 코드 중복 수정, 적절한 상태코드 반환

yoondgu committed:
refactor: URI에 불필요한 path 삭제

yoondgu committed:
refactor: URI 불필요한 path 삭제

yoondgu committed:
refactor: ExceptionHandler 클래스 분리

yoondgu committed:
refactor: Controller 내부 DTO 클래스를 요청 DTO 클래스로 분리

yoondgu committed:
refactor: DDL id에 정수값 unsigned 적용

yoondgu committed:
build: DB 생성 경로, gitignore 수정

yoondgu committed:
refactor: 수정, 삭제 id 검증 로직 추가

- ProductIntegrationTest 매 테스트 케이스마다 초기화 DDL 실행해야 함.
- 그렇지 않으면 테스트 전체 실행 시 수정 성공 케이스 실패함.
- 이를 위해 @Sql 어노테이션 사용함. (다른 방법 확인 필요)

yoondgu committed:
refactor: 서비스 계층 추가, 테스트 작성

- 서비스 테스트는 Stub 방식으로 행위 검증만 진행함.

yoondgu committed:
refactor: Dao 제네릭 적용

yoondgu commented:
안녕하세요 빙봉!! 금방 피드백해주신 것에 비해 리뷰 반영이 좀 늦었네요. ㅠㅠ
연휴 기간에 걸쳐 요청드려서 죄송하단 말씀 드리고 싶어요. 

피드백 및 리팩터링 목록을 [README.md](https://github.com/woowacourse/jwp-shopping-cart/blob/846772307361e6966c5227b3747dd9de24fe003d/README.md)에 남겨주신 코멘트 링크와 함께 정리했습니다!

- 가장 큰 변경은 서비스 계층의 도입, 서비스 테스트 작성입니다.
이 부분은 별도 코멘트로 남긴 것이 없어 설명드리자면,
검증 로직을 추가했지만 여전히 거의 호출만 하는 계층이기는 합니다.
하지만 서비스 테스트 작성을 해볼 수 있다는, 학습 목적으로도 의의가 크다고 생각해 진행했습니다.
서비스 테스트의 경우, Dao 테스트와 달리 "어떤 기능을 호출하고 검증하는지" 확인하는 것이 중요하다고 생각해
Stubbing으로 행위 검증만 하도록 했습니다.
그런데 Mockito도 처음 써보고, 이런 테스트를 작성해본 게 처음이라 제대로 한 것인지 아직 확신이 없습니다. 🥲

- 빈칸으로 남은 것 중 Model vs ModelAndView는 아직 고민이 되어요.
그래서 빙봉은 어떻게 생각하시는지 더 들어보고 싶어서 답변만 남긴 뒤 리뷰 요청 드립니다..! 🤓

감사합니다! 좋은 주말 되세요 🎡

yoondgu committed:
test: displayName 오타 수정

aegis1920 commented on src/main/java/cart/controller/ExceptionAdvice.java:
```diff
@@ -0,0 +1,41 @@
+
+@Slf4j
+@RestControllerAdvice
+public class ExceptionAdvice {
```
- 만약 서비스로직에서 NullPointerException이 발생하면 어떻게 되나요?
- 모든 예외마다 ExceptionHandler를 만들어줘야 할까요? 모든 예외를 처리할 수 있는 ExceptionHandler를 만드려면 어떻게 해야할까요?

yoondgu replied:
모든 예외에 대한 핸들러를 만드는 대신, 예측하지 못한 예외에 대해서도 처리해주도록 하겠습니다!!

aegis1920 commented on src/main/java/cart/controller/ProductController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @PutMapping
```
PUT 메서드를 사용하신 이유가 있을까요? PATCH는 안될까요?

yoondgu replied:
PUT은 자원의 모든 값을 전달받아 업데이트하는 경우, PATCH는 일부 값만 업데이트하는 경우에 사용하는 것으로 알고 있어 PUT 메서드를 사용했습니다!

aegis1920 commented on src/test/java/cart/service/ProductServiceTest.java:
```diff
@@ -0,0 +1,103 @@
+class ProductServiceTest {
+
+    private static final long FIXTURE_ID = 1L;
+    private static final Product FIXTURE_NEW_PRODUCT = new Product("빙봉치킨", "bingbong.jpg", 10000);
+    private static final Product FIXTURE_UPDATED_PRODUCT = new Product(FIXTURE_ID, "도이치킨", "doy.jpg", 10000);
```
😄

aegis1920 commented on src/test/java/cart/service/ProductServiceTest.java:
```diff
@@ -0,0 +1,103 @@
+    private static final Product FIXTURE_NEW_PRODUCT = new Product("빙봉치킨", "bingbong.jpg", 10000);
+    private static final Product FIXTURE_UPDATED_PRODUCT = new Product(FIXTURE_ID, "도이치킨", "doy.jpg", 10000);
+
+    @Mock
```
실제 객체를 사용하지 않고 Mocking을 하신 이유가 있을까요?

yoondgu replied:
"실제 Dao 객체가 잘 동작하는지"에 대한 확인은 DaoTest에서 책임지는 것이라 생각합니다.
따라서 서비스 테스트에서는 Dao 객체가 잘 동작한다는 전제 하에,
특정 기능에서 기대하는 동작을 하는지만 테스트하고 싶었습니다. (상품 추가 -> Dao의 Insert를 호출)

aegis1920 commented on README.md:
```diff
@@ -1 +1,43 @@
+    - [x] 상품 수정 기능 구현
+    - [x] 상품 삭제 기능 구현
+
+## 코드리뷰 피드백 및 리팩터링 목록
```
👍

aegis1920 commented on src/main/java/cart/service/ProductService.java:
```diff
@@ -0,0 +1,46 @@
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional(readOnly = true)
```
- 지금 상황에서 트랜잭션 어노테이션을 사용하신 이유가 있을까요?
- 지금 상황에서 트랜잭션 어노테이션에서 readOnly 옵션을 붙이는 것과 붙이지 않는 것의 차이가 있을까요?

yoondgu replied:
약간 관성적으로(?) 각 메서드가 하나의 트랜잭션으로 취급되어야 한다고 생각해서 붙여주었는데요.
생각해보니 각 메서드에서 모두 단 하나의 DB 작업만 하기 때문에 불필요한 사용이었던 것 같습니다!
현재 필요없는 어노테이션을 붙이면 관리가 어려우니 삭제하도록 하겠습니다.

해당 어노테이션에 readOnly 옵션을 붙이는 이유는, 조회에서는 Lock을 걸어주는 작업이 필요가 없기 때문에 성능 최적화를 위해 많이 사용하는 것으로 알고 있습니다. 
하지만 지금 상황에서도 이러한 성능적 이점이 작용하는지에 대해서는 잘 모르겠습니다. 추측하기로는 트랜잭션 어노테이션을 붙이지 않아도 단일 작업에 대한 트랜잭션 처리가 지원이 되고, 만약 조회만 하는 트랜잭션이라면 알아서 불필요한 Lock은 걸어주지 않을 것 같습니다. 하지만 관련 기반 지식이 아직 부족한지, 더 많이 찾아보고 공부해야 알 것 같아요.
혹시 제 추측이 맞는지, 또는 관련해서 도움이 되는 키워드를 주실 수 있을까요...??! 🥹
(그래도 명시적으로 읽기 전용 메서드임을 동료 개발자에게 알릴 수 있다는 장점은 있을 것 같습니다!!)

aegis1920 replied:
넵! 현재 상황에서는 논리적으로 하나의 단위로 취급될 필요가 없다고 판단했고 트랜잭션 또한 필요없다고 생각되어서 위 질문을 드렸습니다!

> 해당 어노테이션에 readOnly 옵션을 붙이는 이유는, 조회에서는 Lock을 걸어주는 작업이 필요가 없기 때문에 성능 최적화를 위해 많이 사용하는 것으로 알고 있습니다. 하지만 지금 상황에서도 이러한 성능적 이점이 작용하는지에 대해서는 잘 모르겠습니다. 추측하기로는 트랜잭션 어노테이션을 붙이지 않아도 단일 작업에 대한 트랜잭션 처리가 지원이 되고, 만약 조회만 하는 트랜잭션이라면 알아서 불필요한 Lock은 걸어주지 않을 것 같습니다.

- 트랜잭션 어노테이션을 붙이지 않으면 바로 DAO 로직을 타기 때문에 트랜잭션 처리가 지원이 되지 않습니다.
- 트랜잭션 어노테이션을 붙이더라도 어떤 DB를 쓰는지, 어떤 트랜잭션 매니저를 쓰는지에 따라 다릅니다. 지금 같은 상황에서 readOnly 옵션을 붙이더라도 DataSourceTransactionManager.prepareTransactionalConnection()에서 실행되는 isEnforceReadOnly가 false로 되어 `SET TRANSACTION READ ONLY` 구문은 실행되지 않습니다.
  - enforceReadOnly 옵션은 DataSourceTransactionManager을 직접 설정해줘야 설정 가능합니다. (setEnforceReadOnly)
  - 여기 javadoc을 읽어보시면 도움이 될 것 같네요 🙂 
- 결국 지금 상황(h2 인메모리, DataSourceTransactionManager)에서 readOnly 옵션을 붙이든 붙이지 않든 `@Transactional`동작은 같습니다.

yoondgu replied:
빙봉 덕분에 이번 미션에서 내부 동작에 대한 이해의 중요성을 많이 느끼는 것 같아요..!! 감사합니다.
알려주신 내용 기반으로 열심히 학습해보겠습니다 ㅎ.ㅎ

aegis1920 reviewed (approved):
안녕하세요 도이! 빙봉입니다 🙂 

피드백 반영 너무 잘해주셨어요! 👍 
코멘트에 대한 답변 및 추가 코멘트 남겼으니 2단계에서 반영해주시면 될 것 같아요!

1단계는 여기서 머지하겠습니다! 고생하셨어요~
